x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,238 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+					// Sleep for at least 150ms and expect the context passed to server.Connection()
+					// to time out during the sleep. Expect that the context error is ignored
+					// because the dial error is not a timeout.
+					time.Sleep(150 * time.Millisecond)
+					return c, err
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
+
+			// Start a goroutine that pulls events from the events channel and inserts them into the
+			// events slice. Use a sync.WaitGroup to allow the test code to block until the events
+			// channel loop exits, guaranteeing that all events were copied from the channel.
+			events := make([]*event.PoolEvent, 0)
+			eventsCh := make(chan *event.PoolEvent)
+			var eventsWg sync.WaitGroup
+			eventsWg.Add(1)
+			go func() {
+				for evt := range eventsCh {
+					events = append(events, evt)
+				}
+				eventsWg.Done()
+			}()
+
+			// Create a TCP listener on a random port. The listener will accept connections but not
+			// read or write to them.
+			l, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+			defer func() {
+				_ = l.Close()
+			}()
+
+			server, err := NewServer(
+				address.Address(l.Addr().String()),
+				primitive.NewObjectID(),
+				// Create a connection pool event monitor that sends all events to an events channel
+				// so we can assert on the connection pool events later.
+				WithConnectionPoolMonitor(func(_ *event.PoolMonitor) *event.PoolMonitor {
+					return &event.PoolMonitor{
+						Event: func(event *event.PoolEvent) {
+							eventsCh <- event
+						},
+					}
+				}),
+				// Replace the default dialer and handshaker with the test dialer and handshaker, if
+				// present.
+				WithConnectionOptions(func(opts ...ConnectionOption) []ConnectionOption {
+					if tc.dialer != nil {
+						opts = append(opts, WithDialer(tc.dialer))
+					}
+					if tc.handshaker != nil {
+						opts = append(opts, WithHandshaker(tc.handshaker))
+					}
+					return opts
+				}),
+				// Disable monitoring to prevent unrelated failures from the RTT monitor and
+				// heartbeats from unexpectedly clearing the connection pool.
+				withMonitoringDisabled(func(bool) bool { return true }),
+			)
+			require.NoError(t, err)
+			require.NoError(t, server.Connect(nil))
+
+			// Use the context returned by the test case ctxFn to call Connection.
+			ctx, cancel := tc.ctxFn()
+			if cancel != nil {
+				defer cancel()
+			}
+			_, err = server.Connection(ctx)
+			if tc.expectErr {
+				assert.NotNil(t, err, "expected an error but got nil")
+			} else {
+				assert.Nil(t, err, "expected no error but got %s", err)
+			}
+
+			// Close the events channel and expect that no more events are sent on the channel. Then
+			// wait for the events channel loop to return before inspecting the events slice.
+			close(eventsCh)
```

Comment:
IIUC if the `Server` attempts to publish any more events, it will panic because the channel has been closed. Would it be more robust to first call `Server#Disconnect()` and then close the channel?